### PR TITLE
Extend execution context

### PR DESCRIPTION
**Pull request by shargon**
Required for https://github.com/neo-project/neo/issues/927 (https://github.com/neo-project/neo/issues/927#issuecomment-520739728)

The first step for:
https://github.com/neo-project/neo/issues/940

### CODE REVIEW

**Comment by igormcoelho** [comment]
> I manage this with the Clone method, now is virtual

I know you can do this with Clone, but I think perhaps its not a good design.... that's what we call in Brazil, a "gambiarra" :joy:

Can't you use C# superpowers, reflection,something like that to actually create an actual instance?

In fact, do we only need clone (the same) or a similar object?

**Comment by igormcoelho** [comment]
@shargon its worse than I imagined, we need to.remove that factory from LoadScript :joy:

I took a look, only LoadScript creates new ExecutionContext, so we need to put LoadScript as virtual,and rewrite it on extended class.

**Comment by shargon** [comment]
One possible design is with T types, ExecutionEngine<ExecutionContext>

**Comment by igormcoelho** [comment]
can LoadScript be virtual?

We can create an UT and see if its called from other methods. Perhaps pure virtual is necessary... strange feeling.

**Comment by igormcoelho** [comment]
T types is bad for libraries, lets avoid it..

**Comment by igormcoelho** [comment]
I checked, LoadScript can be virtual, not need pure virtual. So we can try your clone solution :joy: :+1:

File: src/neo-vm/ExecutionEngine.cs [hunk 1]
```diff
@@ -1087,13 +1087,9 @@ public virtual ExecutionContext CreateExecutionContext(Script script, Script cal
             return new ExecutionContext(script, callingScript, rvcount);
         }
 
-        public ExecutionContext LoadScript(byte[] script, int rvcount = -1, IExecutionContextFactory factory = null)
+        virtual public LoadScript(byte[] script, int rvcount = -1)
```
Comment (shargon) [inline src/neo-vm/ExecutionEngine.cs]: It must return the ExecutionContext

Comment (igormcoelho) [reply src/neo-vm/ExecutionEngine.cs]: Sorry :joy:

**Comment by igormcoelho** [comment]
@shargon this is fine for me... I think `CreateExecutionContext` can be removed too ( funny thing, I was also thinking on creating such a method hahaha)

In the end, if clone works for CALL instructions, we just need to generate correct instance on LoadScript (which is virtual), and keep cloning it correctly over time.

**Comment by shargon** [comment]
we need LoadContext as protected too

**Comment by igormcoelho** [comment]
> we need LoadContext as protected too

That's true @shargon... inherited method will need to reuse it.

**Comment by igormcoelho** [comment]
Do you think this works now? To help your other proposal?

**Comment by shargon** [comment]
For me, is perfect now

**Comment by erikzhang** [comment]
Why do you need these `virtual` methods?

**Comment by shargon** [comment]
We need it for flag the context as ReadOnly on https://github.com/neo-project/neo/issues/927 , and it will help for caching the scriptHash too

**Comment by igormcoelho** [comment]
On general @erikzhang, the idea is to allow extending the ExecutionContext class, when your inherited execution engine in upper layers will specifics carried on it. 
I think it's very good to reuse neovm as a real library.

**Comment by erikzhang** [comment]
Can we just listen to the `ContextLoaded` event?

**Comment by shargon** [comment]
We can't override nothing in this event, what do you propose?

**Comment by igormcoelho** [comment]
I dont think ContextLoaded will help us Erik.. we need more data on context.

**Comment by erikzhang** [comment]
@shargon @igormcoelho Please check #193

**Comment by shargon** [comment]
Close in favor of https://github.com/neo-project/neo-vm/pull/193

**Pull request closed without merging**